test: Unconditionally check for fRelay field in test framework

jarolrod opened the pull request:
picking up #20411 (rebased onto master)

There is a discrepancy in the implementation of our p2p protocol between
bitcoind and the testing framework.  The fRelay field is an optional
field at the end of a version message as of protocol version 70001.
However, when deserializing a message in bitcoind, we don't check the
version to see if it should have an fRelay field or not.  Instead, we
unconditionally attempt to deserialize into the field.

This commit brings the testing framework in line with the implementation
in core.

This matters for a version message with the following fields:

Version = 60000
fRelay = 1

Bitcoind would deserialize this into a version message with
Version=60000 and fRelay=1, whereas (before this commit) our testing
framework would deserialize this into a version message with
Version=60000 and fRelay=0.

troygiorshev committed:
Unconditionally check for fRelay field in test framework

There is a discrepancy in the implementation of our p2p protocol between
bitcoind and the testing framework.  The fRelay field is an optional
field at the end of a version message as of protocol version 70001.
However, when deserializing a message in bitcoind, we don't check the
version to see if it should have an fRelay field or not.  Instead we
unconditionally attempt to deserialize into the field.

This commit brings the testing framework in line with the implementation
in core.

This matters for a version message with the following fields:

Version = 60000
fRelay = 1

Bitcoind would deserialize this into a version message with
Version=60000 and fRelay=1, whereas (before this commit) our testing
framework would deserialize this into a version message with
Version=60000 and fRelay=0.

jnewbery commented on test/functional/test_framework/messages.py:
```diff
@@ -1045,13 +1045,13 @@ def deserialize(self, f):
+        # But, unconditionally check it to match behaviour in bitcoind
+        try:
+            self.relay = struct.unpack("<b", f.read(1))[0]
+        except KeyboardInterrupt:
+            raise KeyboardInterrupt
+        except:
```
I think better would be to just catch `struct.error`:

```suggestion
        except struct.error:
            self.relay = 0
```

and allow all other exceptions to raise.

jarolrod replied:
addressed in 5c46992

jnewbery replied:
You can remove:

```
        except KeyboardInterrupt:
            raise KeyboardInterrupt
```

since there's no blanket `except` catching, this is doing nothing (`KeyboardInterrupt` exceptions would be raised anyway)

jnewbery reviewed (commented):
Concept ACK. This seems reasonable. fRelay was added with p2p version 70001, but bitcoind has never checked that:

https://github.com/bitcoin/bitcoin/pull/1795/files#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2R2972

I guess that other implementations adopted fRelay without bumping their version number.

jarolrod commented:
Updated from  bfc4333 -> 5c46992, addressed @jnewbery suggestions

Changes:
- catch specific `struct.error` instead of blanket except

jarolrod commented:
Updated from 5c46992 -> 39a9ec5, addressed @jnewbery [comment](https://github.com/bitcoin/bitcoin/pull/21357#discussion_r591467529)

**Changes**: I misunderstood the initial suggestion. This removes the unnecessary handling of `KeyboardInterrupt`

jnewbery commented:
utACK 39a9ec579f023ab262a1abd1f0c869be5b1f3f4d

MarcoFalke commented on test/functional/test_framework/messages.py:
```diff
@@ -1045,13 +1045,11 @@ def deserialize(self, f):
-        else:
+        # Relay field is optional for version 70001 onwards
+        # But, unconditionally check it to match behaviour in bitcoind
+        try:
```
The goal of the test framework is not to match behavior in bitcoind (we don't want to re-implement bitcoind in python), but to properly deserialize messages that have been serialized by bitcoind.

Thus, the try-catch can be removed completely.

jnewbery replied:
The deserialization code can be useful in other contexts, eg https://github.com/bitcoin/bitcoin/blob/f67b5dca576669bfed0cf4af9f2cc6273010a0d7/contrib/message-capture/message-capture-parser.py. Besides, there's no disadvantage to making this more robust.

maflcko reviewed (commented):
seems fine, but I think we don't need any try-catch
